perf: app start time && large table

tiensonqin opened the pull request:
This PR significantly improves the app start time and view performance, especially for large tables, it moves almost all view-related features to the worker, except for custom queries that I plan to address in another PR. It also fixes several issues that result in memory leaks.

### Important changes
1. don't load all pages to the UI db when app starts
2. load less data, only that which the UI needs, e.g. `<get-block` for those blocks that are going to be rendered instead of all the blocks in a view, asking for properties that are need for the UI instead of all the properties
3. render less, e.g. view groups are virtualized, lazy rendering table cells are not in the viewport
4. replace `derived-atom` with missionary flow to avoid memory leak
5. LRU cache for a fewer thread apis
6. linked/unlinked references for file graphs are using views too
7. view `data` are block db ids compared to entities, the UI will call `<get-block` when rendering 
8. [move graph view data handler to deps/db](https://github.com/logseq/logseq/pull/11774/commits/4c54427640fee040b643a5e39ce81fe0eb238f7f)
9. [perf: don't load block children if collapsed](https://github.com/logseq/logseq/pull/11774/commits/005cfeeebb6b285af6f0e5d81d59fdaef9f3bd90)
10. [enhance: add use-flow-state hook](https://github.com/logseq/logseq/pull/11774/commits/5288f171efba947835b551f58a1dda712f825c26)

### Fixed
1. [later picked settings should have higher priority for sorting](https://github.com/logseq/logseq/pull/11774/commits/da742282137fae833636cf93cbd977074d6dcfe7)

### Result
Benchmark graph: 4k movies [Exported sqlite file](https://github.com/logseq/db-benchmark-graphs/blob/master/sqlite/logseq_db_4k%20movies_1744119718.sqlite)

Result compared to `feat/db`:
1. app start time from 20s to 1 ~ 2s
    feat/db: loads 784571 datoms in 20077 ms
    this PR: restore-graph! loads 2880 datoms in 945 ms
2. open "All pages" from 4s to 400ms
3. open "#movies" from 5s to 400ms
4. cmdk search && node reference are much faster, 3~4s to less than 100ms
5. table view scrolling now maintains a minimum frame rate of >40 fps. Previously (in feat/db), such large table views were nearly unusable, with frequent frame drops to 1 fps during scrolling.
6. add/remove object from "#movies" are much faster

RCmerci reviewed (commented):
How much did this PR reduce the app's startup time? Additionally, could you provide the test graph?

tiensonqin commented:
@RCmerci I'm testing with 4k-movies, the app start time was reduced to 2~3s from 16 ~ 18s on my PC, this PR avoids loading all pages (~ 200k datoms) to the memory db when the app starts.

logseq-cldwalker commented:
When clicking into a page with blocks containing a number of properties, the properties aren't loading on initial view. I have to navigate away and return to see them - https://www.loom.com/share/b283f714bfa54998ac802f39f82f7486 . The graph in the video is our [properties graph](https://github.com/logseq/logseq/tree/feat/db/scripts#create-graph-scripts)

logseq-cldwalker commented on deps/db/src/logseq/db.cljs:
```diff
@@ -435,12 +438,30 @@
+        refs (->> (mapcat (fn [id]
+                            (->> (:block/_refs (d/entity db id))
+                                 (remove (fn [ref]
+                                           ;; remove refs that have the block as either tag or property
```
File graphs are no longer showing properties or tags as linked refs. I think this may be the cause

tiensonqin replied:
Fixed by https://github.com/logseq/logseq/commit/43e9f9707c98a44400c768d3a78ba17da3cc9fef.

logseq-cldwalker replied:
Confirm fixed. At some point it would be helpful if logseq.db was more organized with db and file only fns in different namespaces. I wasn't able to notice this by reading the code but only by accident in QA

tiensonqin commented:
> When clicking into a page with blocks containing a number of properties, the properties aren't loading on initial view. I have to navigate away and return to see them - 

It should be fixed by https://github.com/logseq/logseq/commit/2968c84299ed87b2d494a37616b8159067f8e6db.

logseq-cldwalker commented:
> It should be fixed by https://github.com/logseq/logseq/commit/2968c84299ed87b2d494a37616b8159067f8e6db.

Thanks! Also confirmed that I'm no longer seeing any blank pages on db graphs when switching to them

Encountered a file graph regression with block refs:
* Create a block
* In a new block create a reference to first block
* Reference displays as broken instead of displaying first block's title

tiensonqin commented:
@cldwalker the block reference bug should be fixed by https://github.com/logseq/logseq/pull/11774/commits/295ef0f9e1bc3d87e28308e2869403304f25c46d.

logseq-cldwalker commented:
> @cldwalker the block reference bug should be fixed by [295ef0f](https://github.com/logseq/logseq/commit/295ef0f9e1bc3d87e28308e2869403304f25c46d).

Thanks!

A bug I'm seeing with adding a page property:
- Create a page
- Click on 'Set property' for the page and select Description
- Sometimes it pops up another property select and sometimes it does nothing

tiensonqin commented:
> A bug I'm seeing with adding a page property:

I think https://github.com/logseq/logseq/commit/8acdf7563b5b398cbe652dc801407f901b264e73 has fixed the bug.

logseq-cldwalker commented:
> I think https://github.com/logseq/logseq/commit/8acdf7563b5b398cbe652dc801407f901b264e73 has fixed the bug.

Yep. I'm about to log a few more bugs

tiensonqin commented:
@logseq-cldwalker No problem, I plan to merge this PR next week so that we have more time to test.

logseq-cldwalker commented:
Some bugs I've found during QA:

1. In db graphs, advanced query block heights look way too tall:

<img width="370" alt="Screen Shot 2025-04-10 at 10 48 19 AM" src="https://github.com/user-attachments/assets/c074b7e7-040e-4471-ac40-670e0b459bd3" />

2. In file graphs, in journals screen, typing in a block of the 2nd journal page gets deleted - https://www.loom.com/share/a1f8e6866a1a4a9882678b7c8acb2b09

3. In file graphs, linked references do not show pages for the blocks. The example below is linked refs for the p1 property page

<img width="370" alt="Screen Shot 2025-04-10 at 9 57 11 PM" src="https://github.com/user-attachments/assets/9cd3b519-1d5a-46c9-9da7-332d069a2c28" />

4. In file graphs, page names for query results show up as `(())`` instead of showing the page names. This also happens with the page names in the home page queries e.g. scheduled and deadline - https://www.loom.com/share/3046bcbf65594aa49f0174a3a5b0e850 . The file graph in the video is https://github.com/logseq/logseq/tree/perf/app-start/deps/graph-parser/test/resources/exporter-test-graph

logseq-cldwalker reviewed (commented):
Nice work. Fixed a few small things, code looks great and found [the following bugs](https://github.com/logseq/logseq/pull/11774#issuecomment-2795804514). I just have some QA of large graphs to do on Monday

tiensonqin commented:
> In db graphs, advanced query and #Code blocks height look way too tall:
> In file graphs, linked references do not group blocks by page or show page titles. The example below is linked refs for the p1 property page
> Creating views in All pages, class or property tables, result in blank views that persist on that page until you navigate away and come back -
> Property descriptions do not load on for the properties on a class' page -

Those bugs should be fixed now.

> In file graphs, in journals screen, typing in a block of the 2nd journal page gets deleted 
> In file graphs, page names for query results show up as `(())`` instead of showing the page names. This also happens with the page names in the home page queries, e.g. scheduled and deadline

I failed to reproduce those two bugs, let me know if they still don't work.

logseq-cldwalker commented:
> Those bugs should be fixed now.

Confirmed all fixed although the creating view has a different bug - https://www.loom.com/share/387036370a154e568dccb590742abe34. I'm ok with filing this to be fixed later. Repro steps:
* Create a new view in a tag or property page
* In the new view, filter down the results
* Switch back to the All view and it shows the nodes from the new view instead of all nodes
* Navigating away from the screen and back fixes the caching

> In file graphs, in journals screen, typing in a block of the 2nd journal page gets deleted

This is fixed. The re-rendering that was deleting text no longer happens

> In file graphs, page names for query results show up as `(())`` instead of showing the page names. This also happens with the page names in the home page queries, e.g. scheduled and deadline

This is still happening. Repro steps:
* On desktop, open this file graph as a file graph - https://github.com/logseq/logseq/tree/perf/app-start/deps/graph-parser/test/resources/exporter-test-graph
* On the latest journal page, type `{{query (property prop-num)}}` in a new block
* The page titles do not load until navigating to another screen like All pages
* If you switch to another graph and back, the bug reappears


Some new regressions I've encountered:
1. The filter for linked references doesn't take affect until you navigate away. Affects both graph types
2. Pressing enter in in code and advanced query blocks no longer creates a newline
3. In file graphs, pressing enter on the tag autocomplete results in an error most of the time

I can take a look at the last bug after I finish QA

tiensonqin commented:
> In file graphs, pressing enter on the tag autocomplete results in an error most of the time

@logseq-cldwalker Can you record a video? I'm unable to reproduce it.

tiensonqin commented:
I think all the bugs except the `enter on the tag autocomplete` one have been fixed.

tiensonqin commented:
Not sure if it's related, there's a page that's missing some attributes in the graph `exporter-test-graph`:
```
{:block/name "some ", :db/id 354}
```

logseq-cldwalker commented:
> I think all the bugs except the enter on the tag autocomplete one have been fixed.

Thanks! They've all been fixed. `enter on the tag autocomplete` went away once I re-indexed the file graph

> there's a page that's missing some attributes in the graph exporter-test-graph

Is this as a file graph or db graph imported from the file graph? I'm not seeing nil :block/title's for pages as a file graph

logseq-cldwalker reviewed (approved):
@tiensonqin Thanks to you and @RCmerci for all the improvements here! 👍 💪  I QAed the mentioned perf improvements for 2 large graphs. All seemed to work except for:

> this PR: restore-graph! loads 2880 datoms in 945 ms

I'm not seeing this for desktop or browser. Seeing the following on startup and when switching to the movies-4k graph: `:frontend.db.restore/restore-graph! loads 1187104 datoms in 16621 ms`

One optional regression I'm seeing with file graphs: In file graphs, `#` autocomplete also includes blocks, not just pages

tiensonqin commented:
> I'm not seeing this for desktop or browser. Seeing the following on startup and when switching to the movies-4k graph

Fixed by https://github.com/logseq/logseq/pull/11774/commits/86d2adc4f005d326048e090b45d9b2a4ffe230b7

> One optional regression I'm seeing with file graphs: In file graphs, # autocomplete also includes blocks, not just pages

Fixed by https://github.com/logseq/logseq/pull/11774/commits/0105ee53d29bfed641a3a7196dfb27c389275197
